### PR TITLE
ROS2 CI dev

### PR DESCRIPTION
**Pull request by niiquaye**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,91 +1,32 @@
...
 on:
-  push:
   pull_request:
-  schedule: # Trigger a job on master at midnight UTC every day
-    - cron: '0 0 * * *'
+  push:
+    branches:
+      - ros2
+      - master
+      - ci-dev
```
Comment (wmmc88) [inline .github/workflows/ci.yml:9]:
```suggestion
      - ros2
```

This should only list ros2 for now.

when we merge ros2 into master, we'll replace ros2 here with master**

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,91 +1,32 @@
 name: CI
 
 on:
-  push:
   pull_request:
-  schedule: # Trigger a job on master at midnight UTC every day
-    - cron: '0 0 * * *'
```
Comment (wmmc88) [inline .github/workflows/ci.yml:7]: You wan to keep these lines here. They:
1. run pipelines on every pushed commit
2. run pipelines on every pr (iirc they run on a merged reference but dont quote me on that)
3. run pipelines on the master every day at midnight to make sure that 3rd party package updates don't introduce breaking changes for us

Comment (wmmc88) [reply .github/workflows/ci.yml:7]: we probs actually want to change 3 to run on both master and ros2 for now

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,91 +1,32 @@
...
+      with:
+        target-ros2-distro: foxy
+        package-name: uwrt_mars_rover
 
-      - name: Install Qemu and Start Qemu Docker
-        shell: bash
-        run: |
-          sudo apt update
-          sudo apt install qemu-user-static
-          docker run --rm --privileged multiarch/qemu-user-static --reset --credential yes --persistent yes
```
Comment (wmmc88) [inline .github/workflows/ci.yml:87]: you're probably gonna need to add this back somewhere to get arm64 builds working

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1,91 +1,112 @@
...
-        uses: ros-industrial/industrial_ci@master
-        env: ${{matrix.env}}
-
-  clang-format:
+    - name: Set Environment Variables 
+      run: |
+        echo "name=LANG" >> $GITHUB_ENV
+
+    - name: Setup ROS 2
+      uses: ros-tooling/setup-ros@0.1.4
```
Comment (wmmc88) [inline .github/workflows/ci.yml:30]: This should probably use the docker version of this action as it'll be faster

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -1,91 +1,112 @@
...
+      uses: ros-tooling/action-ros-ci@0.2.0
+      with:
+        target-ros2-distro: foxy
+        uncrustify: false
+        package-name: uwrt_mars_rover
+
+  # ament_lint job will run the following linters:
+  #  - flake8 (python linter)
+  #  - clang-format (clang format check - Note: this runs clang-format-10)
+  #  Note: this job is much faster than the above job because the use of a docker container
```
Comment (wmmc88) [inline .github/workflows/ci.yml:44]: The speed difference here is actually because you arent building the code at all in this step. flake8 and clang-format dont require the code to be built to run

Comment (wmmc88) [reply .github/workflows/ci.yml:44]: From the `action-ros-lint` readme:

> This action runs ROS 2 linters on a ROS 2 package. This command does not compile any code intentionally to make it as fast as possible (<2 min). The objective is to give contributors feedback about their change quickly. It may also be used to avoid wasting CI resources (no need to compile, and run the tests if it does not pass the linters).

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -1,91 +1,112 @@
...
+    - name: Set Environment Variables 
+      run: |
+        echo "name=LANG" >> $GITHUB_ENV
+
+    - name: Setup ROS 2
+      uses: ros-tooling/setup-ros@0.1.4
+      with:
+        required-ros-distributions: foxy
+    
+    - name: Run Tests
```
Comment (wmmc88) [inline .github/workflows/ci.yml:34]: you should probably run `ros-tooling/action-ros-lint` here before building so that it doesnt bother to waste build resources on this job if it fails the lint checks.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -1,91 +1,112 @@
...
+              "build":{
+                "cmake-args":[
+                    "-DCMAKE_EXPORT_COMPILE_COMMANDS=ON" 
+                ]
+
+              }
+            }
+      - uses: ros-tooling/action-ros-lint@master
+        with:
+          distribution: rolling
```
Comment (wmmc88) [inline .github/workflows/ci.yml:106]: is there a reason you're using rolling here?

File: uwrt_mars_rover/CMakeLists.txt [hunk 8]
```diff
@@ -1,4 +1,35 @@
...
+
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  # the following line skips the linter which checks for copyrights
+  # uncomment the line when a copyright and license is not present in all source files
+  #set(ament_cmake_copyright_FOUND TRUE)
+  # the following line skips cpplint (only works in a git repo)
+  # uncomment the line when this package is not in a git repo
+  #set(ament_cmake_cpplint_FOUND TRUE)
+  ament_lint_auto_find_test_dependencies()
```
Comment (wmmc88) [inline uwrt_mars_rover/CMakeLists.txt:32]: The tests that get run here are based off the test dependencies specified in the xml. The proper way to disable running thise tests is just not to include the `ament_lint_common` test depend and include each linter individually there

Comment (wmmc88) [reply uwrt_mars_rover/CMakeLists.txt:32]: https://github.com/ament/ament_lint/blob/master/ament_lint_common/doc/index.rst

**Review by wmmc88 — CHANGES_REQUESTED** [review]
Additional blockers for this pr i think are too enable code coverage reporting( ie. lcov for c++, coveragepy  for python, and codecov) and also enable asan testing. 

It should also implement the custom rosinstall files: https://github.com/ros-tooling/action-ros-ci#build-with-a-custom-repos-or-rosinstall-file

File: uwrt_mars_rover/CMakeLists.txt (hunk 8 above)
Comment (niiquaye) [reply uwrt_mars_rover/CMakeLists.txt:32]: > https://github.com/ament/ament_lint/blob/master/ament_lint_common/doc/index.rst



> The tests that get run here are based off the test dependencies specified in the xml. The proper way to disable running thise tests is just not to include the `ament_lint_common` test depend and include each linter individually there

Oh nice I will do that then

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -1,91 +1,70 @@
 name: CI
 
 on:
-  push:
   pull_request:
-  schedule: # Trigger a job on master at midnight UTC every day
+  push:
```
Comment (niiquaye) [inline .github/workflows/ci.yml:5]: swap this

**Comment by wmmc88** [comment]
I think you want to remove the "ROS2 Drivetrain Description" commit or squash the last two commits together

File: .clang-format [hunk 10]
```diff
@@ -3,4 +3,4 @@ BasedOnStyle: Google
 ColumnLimit: 120
 AllowShortFunctionsOnASingleLine: Empty
 CommentPragmas: '^ NOLINT'
----
```
Comment (wmmc88) [inline .clang-format:6]: Why was this removed? i think it might be needed

File: uwrt_mars_rover/CMakeLists.txt [hunk 11]
```diff
@@ -1,4 +1,60 @@
-cmake_minimum_required(VERSION 2.8.3)
+cmake_minimum_required(VERSION 3.5)
```
Comment (wmmc88) [inline uwrt_mars_rover/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.14.4)
```

Should change to match the min required version needed for required platforms as defined in REP 2000: https://www.ros.org/reps/rep-2000.html#foxy-fitzroy-may-2020-may-2023

File: uwrt_mars_rover/CMakeLists.txt [hunk 12]
```diff
@@ -1,4 +1,60 @@
...
+if (CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif ()
+
+# find dependencies
+find_package(ament_cmake REQUIRED)
+find_package(rclcpp REQUIRED)
+# uncomment the following section in order to fill in
+# further dependencies manually.
+# find_package(<dependency> REQUIRED)
```
Comment (wmmc88) [inline uwrt_mars_rover/CMakeLists.txt:23]:
```suggestion
```

File: uwrt_mars_rover/CMakeLists.txt [hunk 13]
```diff
@@ -1,4 +1,60 @@
...
+  find_package(ament_cmake_xmllint REQUIRED)
+  ament_xmllint()
+
+  # clang-tidy
+  find_package(ament_cmake_clang_tidy REQUIRED)
+  ament_clang_tidy(${CMAKE_BINARY_DIR})
+
+  # the following line skips cpplint (only works in a git repo)
+  # uncomment the line when this package is not in a git repo
+  #set(ament_cmake_cpplint_FOUND TRUE)
```
Comment (wmmc88) [inline uwrt_mars_rover/CMakeLists.txt:56]: should this still be commented?

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -3,89 +3,110 @@ name: CI
...
+        # todo foxy
+
+      - name: Build ROS2 Packages and Run Tests
+        run: >
+          docker exec arm64-ros bash -c "source /opt/ros/foxy/setup.bash;
+          colcon build --event-handlers console_cohesion+;
+          colcon test --event-handlers console_cohesion+"
+        # --packages-up-to
+
+# todo: distro matrix, windows build, global packages list
```
Comment (wmmc88) [inline .github/workflows/ci.yml:112]: i've got some todos here that need to be finished still. if u really wanna merge this in, i can fix em in a follow up pr.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -3,89 +3,110 @@ name: CI
...
+        run: |
+          docker exec arm64-ros sudo apt update
+          docker exec arm64-ros rosdep install -r --from-paths . --ignore-src --rosdistro foxy -y
+        # todo foxy
+
+      - name: Build ROS2 Packages and Run Tests
+        run: >
+          docker exec arm64-ros bash -c "source /opt/ros/foxy/setup.bash;
+          colcon build --event-handlers console_cohesion+;
+          colcon test --event-handlers console_cohesion+"
```
Comment (wmmc88) [inline .github/workflows/ci.yml:109]:
```suggestion
        run: >
          docker exec arm64-ros bash -c
          "source /opt/ros/foxy/setup.bash;
          colcon build --event-handlers console_cohesion+;
          colcon test --event-handlers console_cohesion+"
```

fyi using `>` instead of `|` folds newlines into spaces

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -3,89 +3,110 @@ name: CI
 on:
```
Comment (wmmc88) [inline .github/workflows/ci.yml:3]: should probably add the cancel workflow action to cancel jobs on commits that have been superceeded in order to save build resources. I set it up in fw so can take a look there. could be a follow up thing

**Comment by wmmc88** [comment]
Woah @younesr1 r ya taking over the pr?

**Comment by younesr1** [comment]
> Woah @younesr1 r ya taking over the pr?

nah lol just hadnt realised i was on ci-dev

File: .github/workflows/ci.yml (hunk 16 above)
Comment (wmmc88) [reply .github/workflows/ci.yml:3]: done

File: uwrt_mars_rover/CMakeLists.txt (hunk 11 above)
Comment (wmmc88) [reply uwrt_mars_rover/CMakeLists.txt:1]: done

File: .github/workflows/ci.yml (hunk 14 above)
Comment (wmmc88) [reply .github/workflows/ci.yml]: will follow up on these later
@niiquaye

**Review by wmmc88 — APPROVED** [review]
